### PR TITLE
many: ensure-dir mounts for personal-files missing dirs

### PR DESCRIPTION
**Pull request by ernestl**
Generate new ensure-dir kind user mount entries in snap.<snap-name>.user-fstab from personal-files interface write attributes to instruct snap-update-ns to create potentially missing parent directories and provide `snap-update-ns` the minimum required permissions to create the potentially missing parent directories as per `snap.<snap-name>.user-fstab` on connection of the personal-files interface.

This PR `part 2` of a larger feature to automatically create missing personal-files interface parent directories.

There are 3 parts in total:
- Part 1: `pass env real-home and uid to snap-update-ns`              | [pull/13244](https://github.com/snapcore/snapd/pull/13244)
- Part 2: `many: ensure-dir mounts for personal-files missing dirs` | [pull/13260](https://github.com/snapcore/snapd/pull/13260)
- Part 3: `add snap-update-ns support for ensure-dir mounts`        | [pull/13342](https://github.com/snapcore/snapd/pull/13342)

Additional information:
- [Relevant section in spec](https://docs.google.com/document/d/1yRqiFcdAIUjwakIQdijtW61vKojcJN2vTjrcwJvLAdo/edit#heading=h.2k4cf7kofeho)
- Jira [SNAPDENG-8233](https://warthogs.atlassian.net/browse/SNAPDENG-8233)

[SNAPDENG-8233]: https://warthogs.atlassian.net/browse/SNAPDENG-8233?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: strutil/pathiter.go [hunk 1]
```diff
@@ -82,6 +82,11 @@ func (iter *PathIterator) CurrentPath() string {
 	return iter.path[:iter.right]
 }
 
+// CurrentCleanPath returns the same value as CurrentPath with the right slash trimmed.
+func (iter *PathIterator) CurrentCleanPath() string {
```
Comment (ernestl) [inline strutil/pathiter.go:86]: This is also used in `Part 4` for the new variation of MkdirAllWithin.

**Comment by alexmurray** [comment]
It is hard to review these various PRs as they each contain parts of each other from what I can see - which commits from this PR should be reviewed in this case?

**Comment by ernestl** [comment]
> It is hard to review these various PRs as they each contain parts of each other from what I can see - which commits from this PR should be reviewed in this case?

Hi @alexmurray, yes apologies. Ideally a systematic review from part 1 to part 5 where independant working parts are landed would clear things up on the go, but I do not think this will be the case in practice,

For how I indicated exactly which commits are relevant to each PR.

**Review by MiguelPires — APPROVED** [review]
LGTM

File: interfaces/builtin/personal_files.go [hunk 2]
```diff
@@ -139,3 +140,23 @@ func (iface *personalFilesInterface) MountConnectedPlug(spec *mount.Specificatio
...
 	return nil
 }
+
+func (iface *personalFilesInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	if err := iface.commonFilesInterface.AppArmorConnectedPlug(spec, plug, slot); err != nil {
+		return err
+	}
+
+	var writes []interface{}
+	_ = plug.Attr("write", &writes)
```
Comment (pedronis) [inline interfaces/builtin/personal_files.go:150]: same comment as in the previous PR

File: interfaces/mount/spec.go [hunk 3]
```diff
@@ -150,6 +150,29 @@ func (spec *Specification) AddExtraLayouts(layouts []snap.Layout) {
 	}
 }
 
+// createEnsureDirMountEntry creates a mount entry from the given ensure directory spec
+// that instruct snap-update-ns to create missing directories if required.
+func createEnsureDirMountEntry(ensureDirSpec *interfaces.EnsureDirSpec) osutil.MountEntry {
```
Comment (pedronis) [inline interfaces/mount/spec.go:155]: the apparmor code is cleaning the dirs but this code is not? should they be consistent?

File: interfaces/apparmor/spec_test.go [hunk 4]
```diff
@@ -537,6 +537,26 @@ func (s *specSuite) TestApparmorExtraLayouts(c *C) {
...
 }
 
+func (s *specSuite) TestAllowEnsureDirMounts(c *C) {
+	ensureDirSpecs := []*interfaces.EnsureDirSpec{
+		{MustExistDir: "$HOME", EnsureDir: "$HOME/.local/share"},
+		{MustExistDir: "$HOME", EnsureDir: "$HOME/dir1/dir2"},
+		{MustExistDir: "/", EnsureDir: "/dir1/dir2"},
+		{MustExistDir: "/dir1", EnsureDir: "/dir1"},
+	}
+	s.spec.AllowUserEnsureDirMounts("personal-files", ensureDirSpecs)
```
Comment (pedronis) [inline interfaces/apparmor/spec_test.go:547]: it's a bit strange that a function with user in the name let us manipulate / ? maybe it should error or ignore on things that don't have MustExistDIr == $HOME ?

File: interfaces/apparmor/spec.go [hunk 5]
```diff
@@ -567,6 +567,61 @@ func snippetFromLayout(layout *snap.Layout) string {
...
+	}
+	emit := spec.AddUpdateNSf
+
+	// Create entry for MustExistDir
+	iter, err := strutil.NewPathIterator(ensureDir)
+	if err != nil {
+		panic(err)
+	}
+	for iter.Next() {
+		if iter.CurrentCleanPath() == mustExistDir {
```
Comment (pedronis) [inline interfaces/apparmor/spec.go:599]: why do we need a Clean variant given that we clean the inputs above?

**Review by pedronis — COMMENTED** [review]
some questions

File: interfaces/apparmor/spec.go (hunk 5 above)
Comment (ernestl) [reply interfaces/apparmor/spec.go:602]: Yes, `mustExistDir` will be clean, and the iterator will output `<mustExistDir>/`

File: interfaces/builtin/personal_files.go (hunk 2 above)
Comment (ernestl) [reply interfaces/builtin/personal_files.go]: Done

File: interfaces/mount/spec.go (hunk 3 above)
Comment (ernestl) [reply interfaces/mount/spec.go:155]: I have now removed the cleaning from apparmor counterpart. 

The BeforePreparePlug checks covers everything including clean paths and prefix $HOME etc.
The existing pattern seems to be to trust such checks and not re-validate structures passed, thus typically avoiding errors mid emits.

I suppose it is not ideal for spec functions to rely that all interfaces construct its inputs correctly, they should arguably be more independent and have a catch all. 

Let me know your preference, but for now i think this follows the existing pattern.

File: interfaces/apparmor/spec_test.go (hunk 4 above)
Comment (ernestl) [reply interfaces/apparmor/spec_test.go]: I started with a generic function that covers both user and system use of ensure-dir. Then later tried to indicate the currently intended scope by adding ...User...

Thinking about it again, I see no reason why it cannot be a generic function. I added additional comments to show where additional expansions should be added on a need to have basis.

File: interfaces/apparmor/spec.go (hunk 5 above)
Comment (pedronis) [reply interfaces/apparmor/spec.go:602]: but ensureDir is also clean, no? I'm still unsure why the need for CurrentCleanPath ?

File: interfaces/apparmor/spec.go [hunk 6]
```diff
@@ -567,6 +567,64 @@ func snippetFromLayout(layout *snap.Layout) string {
...
+			path = path + "/"
+		}
+		return path
+	}
+	emit := spec.AddUpdateNSf
+
+	// Create entry for MustExistDir
+	iter, err := strutil.NewPathIterator(ensureDir)
+	if err != nil {
+		panic(err)
```
Comment (pedronis) [inline interfaces/apparmor/spec.go:599]: why the panic?  isn't it better to propagate an internal error?

File: interfaces/apparmor/spec.go (hunk 5 above)
Comment (pedronis) [reply interfaces/apparmor/spec.go:602]: actually NewPathIterator double checks that, so I'm still not sure why this change to PathIterator

**Review by pedronis — COMMENTED** [review]
more questions

File: strutil/pathiter.go (hunk 1 above)
Comment (pedronis) [reply strutil/pathiter.go:86]: maybe calling it CurrentPathNoSlash would be clearer?

File: interfaces/apparmor/spec.go (hunk 5 above)
Comment (pedronis) [reply interfaces/apparmor/spec.go:602]: now I see, I think I had the wrong mental model of what PathIterator does, anway a different name would be better

File: interfaces/apparmor/spec.go [hunk 7]
```diff
@@ -567,6 +567,64 @@ func snippetFromLayout(layout *snap.Layout) string {
 	return fmt.Sprintf("# Layout path: %s\n# (no extra permissions required for symlink)", mountPoint)
 }
 
+// emitEnsureDir creates an apparmor snippet that permits snap-update-ns to create
+// missing directories for the calling user according to the provided ensure directory spec.
+// This function is currently used as counterpart for AddUserEnsureDirs, but can also be used
+// for permitting non-user ensure directory specs.
+func emitEnsureDir(spec *Specification, ifaceName string, ensureDirSpec *interfaces.EnsureDirSpec) {
+	ensureDir := ensureDirSpec.EnsureDir
+	mustExistDir := ensureDirSpec.MustExistDir
```
Comment (pedronis) [inline interfaces/apparmor/spec.go:576]: should we check that mustExistDir is a prefix of ensureDir ?

**Review by pedronis — COMMENTED** [review]
I understood things a bit better

File: interfaces/apparmor/spec.go (hunk 5 above)
Comment (ernestl) [reply interfaces/apparmor/spec.go:602]: Apologies if I am missing something.

The problem in my view is this:
Screenshot from 2023-10-26 14-47-27

Only the full path is clean (does not have final slash), but for all directories except the full path, it returns path with "/".

File: interfaces/apparmor/spec.go (hunk 7 above)
Comment (ernestl) [reply interfaces/apparmor/spec.go:576]: As discussed, we do error checking on mount side. This algo already [iterates to after `mustExistDir`](https://github.com/snapcore/snapd/pull/13260/files#diff-78692e23a83df81d72626c6a2a4cb7d54281f8fbaa33073bf019555fe760500aR602), so will just not add anything in this case.

File: interfaces/apparmor/spec.go (hunk 6 above)
Comment (ernestl) [reply interfaces/apparmor/spec.go:599]: yes

File: strutil/pathiter.go (hunk 1 above)
Comment (ernestl) [reply strutil/pathiter.go]: I changed all pathiter method that "clean" to ...NoSlash

File: interfaces/apparmor/spec_test.go (hunk 4 above)
Comment (ernestl) [reply interfaces/apparmor/spec_test.go]: As discussed: 
- avoid errors on the apparmor spec emit side
- check for errors on the mount spec side

Proposal: I added a definitive validation function for ensure-dir specs, that is used on the mount side when the interface attempts to add mount entries on connected plug. It result in an internal error if for whatever reason the interface creating the ensire-dir specs breaks the rules. If this is too much, I am happy to remove.

File: osutil/mountentry.go [hunk 8]
```diff
@@ -132,6 +132,28 @@ func (e *MountEntry) OptStr(name string) (string, bool) {
 	return "", false
 }
 
+// ReplaceMountEntryOption replaces the first mount entry option that has
+// the same prefix as the given option with the given option
+func ReplaceMountEntryOption(entry *MountEntry, option string) {
```
Comment (ernestl) [inline osutil/mountentry.go:137]: Alternative replacement function as requested

**Review by alexmurray — COMMENTED** [review]
Apologies, I still need to spend more time on this one but wanted to point out one issue I noticed so far.

File: interfaces/apparmor/spec.go [hunk 9]
```diff
@@ -567,6 +567,64 @@ func snippetFromLayout(layout *snap.Layout) string {
...
+
+	// Add additional expansion here as required
+	replacePrefixHome := func(path string) string {
+		if strings.HasPrefix(path, "$HOME") {
+			return strings.Replace(path, "$HOME", "@{HOME}", -1)
+		}
+		return path
+	}
+	appArmorDir := func(path string) string {
+		if path != "/" {
```
[the remark this answers is not in the document]
Comment (ernestl) [reply interfaces/apparmor/spec.go:589]: The function adds a slash to a path without slash (CurrentPathNoSlash), in other words <dir-name>/.
The exception is if path is root "/", which [is what is tested for here](https://github.com/snapcore/snapd/pull/13260/files#diff-a62c473a598df9d999f2138f431cebb1c8b58639b42bbdd596b646ac52c34ff7R544). So your proposal will also work, but arguably the existing check better demonstrates the intent?

[the remark this answers is not in the document]
Comment (alexmurray) [reply interfaces/apparmor/spec.go:589]: ah thanks for the clarification - I missed that detail.

File: interfaces/apparmor/spec.go (hunk 6 above)
Comment (ernestl) [reply interfaces/apparmor/spec.go]: Thinking about the again, I changed it to not panic and rather not emit anything, which seems consistent with the concept of validating on the mount side and ignoring errors on the apparmor side.

**Review by pedronis — CHANGES_REQUESTED** [review]
thanks, did another pass, some small comments

File: interfaces/builtin/utils.go [hunk 10]
```diff
@@ -180,3 +181,15 @@ func getDesktopFileRules(snapInstanceName string) []string {
 
 	return rules
 }
+
+// StringListAttribute returns a list of strings for the given attribute key if the attribute exists.
+func StringListAttribute(attrer interfaces.Attrer, key string) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (ernestl) [reply interfaces/builtin/utils.go]: Fixed

File: interfaces/ensure_dir.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+// the remainder of missing directories of EnsureDir.
+type EnsureDirSpec struct {
+	MustExistDir string
+	EnsureDir    string
+}
+
+// IsValid returns an error if the ensure directory specification is not valid.
+func (spec *EnsureDirSpec) Validate() error {
+	if spec.MustExistDir != filepath.Clean(spec.MustExistDir) {
+		return fmt.Errorf("Directory that must exist %q is not a clean path", spec.MustExistDir)
```
[the remark this answers is not in the document]
Comment (ernestl) [reply interfaces/ensure_dir.go]: Fixed everywhere

File: interfaces/apparmor/spec.go [hunk 12]
```diff
@@ -567,6 +567,64 @@ func snippetFromLayout(layout *snap.Layout) string {
...
+
+	// Create entries for the remaining directories after MustExistDir up to and including EnsureDir
+	for iter.Next() {
+		emit("  owner %s/ rw,", replacePrefixHome(iter.CurrentPathNoSlash()))
+	}
+}
+
+// AllowEnsureDirMounts adds snap-update-ns snippets that permit snap-update-ns to create
+// missing directories according to the provided ensure directory mount specs.
+func (spec *Specification) AllowEnsureDirMounts(ifaceName string, ensureDirSpecs []*interfaces.EnsureDirSpec) {
```
[the remark this answers is not in the document]
Comment (ernestl) [reply interfaces/apparmor/spec.go]: Done

File: interfaces/ensure_dir.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+// EnsureDirSpec contains the information required to ensure the existence of a directory.
+// MustExistDir is the prefix of EnsureDir and must exist as prerequisite for creation for
+// the remainder of missing directories of EnsureDir.
+type EnsureDirSpec struct {
+	MustExistDir string
+	EnsureDir    string
+}
+
+// IsValid returns an error if the ensure directory specification is not valid.
```
[the remark this answers is not in the document]
Comment (ernestl) [reply interfaces/ensure_dir.go]: Done

File: interfaces/apparmor/spec.go [hunk 14]
```diff
@@ -567,6 +567,64 @@ func snippetFromLayout(layout *snap.Layout) string {
...
+		return path
+	}
+	emit := spec.AddUpdateNSf
+
+	// Create entry for MustExistDir
+	iter, err := strutil.NewPathIterator(ensureDir)
+	if err != nil {
+		return
+	}
+	for iter.Next() {
```
[the remark this answers is not in the document]
Comment (ernestl) [reply interfaces/apparmor/spec.go:601]: This was discussed with Sameule, more broadly we decided to only raise errors on the mount side of things, and on the apparmor side just not emit if there are unexpected issues: https://github.com/snapcore/snapd/pull/13260#discussion_r1374125255

**Review by pedronis — APPROVED** [review]
thank you

File: interfaces/builtin/personal_files.go [hunk 15]
```diff
@@ -76,3 +81,78 @@ func init() {
...
+	// Create missing directories for write paths only.
+	// BeforePreparePlug should prevent error.
+	writes, _ := stringListAttribute(plug, "write")
+	ensureDirSpecs, err := dirsToEnsure(writes)
+	if err != nil {
+		return fmt.Errorf("cannot connect plug %s: %v", plug.Name(), err)
+	}
+	if len(ensureDirSpecs) > 0 {
+		if err = spec.AddUserEnsureDirs(ensureDirSpecs); err != nil {
+			return fmt.Errorf("cannot connect plug %s, %v", plug.Name(), err)
```
Comment (MiguelPires) [inline interfaces/builtin/personal_files.go:133]:
```suggestion
			return fmt.Errorf("cannot connect plug %s: %v", plug.Name(), err)
```

**Review by MiguelPires — APPROVED** [review]

**Pull request merged**